docs/notebooks/montecarlo.pct.py

```diff
@@ -0,0 +1,177 @@
+)
+from util.plotting import plot_function_2d, plot_bo_points, plot_regret
+
+# %%
+gpflow.config.set_default_float(np.float64)
+np.random.seed(1793)
+tf.random.set_seed(1793)
+
+# %% [markdown]
+# ## Describe the problem
+# In this example, we look to find the minimum value of the two-dimensional Branin function over the hypercube $[0, 1]^2$. We can plot contours of the Branin over this space.
+
+# %%
+mins = [0.0, 0.0]
+maxs = [1.0, 1.0]
+
+fig = plot_function_plotly(branin, mins, maxs, grid_density=20)
+fig.update_layout(height=400, width=400)
+fig.show()
+
+# %% [markdown]
+# ## Sample the observer over the search space
+#
+# Sometimes we don't have direct access to the objective function. We only have an observer that indirectly observes it. In _Trieste_, the observer outputs a number of datasets, each of which must be labelled so the optimization process knows which is which. In our case, we only have one dataset, the objective. We'll use _Trieste_'s default label for single-model setups, `OBJECTIVE`. We can convert a function with `branin`'s signature to a single-output observer using `mk_observer`.
+#
+# The optimization procedure will benefit from having some starting data from the objective function to base its search on. We sample five points from the search space and evaluate them on the observer. We can represent the search space using a `Box`.
+
+# %%
+observer = mk_observer(branin, OBJECTIVE)
+lower_bound = tf.cast(mins, gpflow.default_float())
+upper_bound = tf.cast(maxs, gpflow.default_float())
+search_space = trieste.space.Box(lower_bound, upper_bound)
+
+num_initial_points = 10
+initial_query_points = search_space.sample(num_initial_points)
+initial_data = observer(initial_query_points)
+
+# %% [markdown]
+# ## Model the objective function
+#
+# The Bayesian optimization procedure estimates the next best points to query by using a probabilistic model of the objective. We'll use Gaussian process regression for this, provided by GPflow. The model will need to be trained on each step as more points are evaluated, so we'll package it with GPflow's Scipy optimizer.
+#
+# Note we could leave it to the optimizer to build this model, but we'll want to inspect it later so we'll build it ourselves.
+#
+# Just like the data output by the observer, the optimization process assumes multiple models, so we'll need to label the model in the same way.
+
+# %%
+def build_model(data):
+    variance = tf.math.reduce_variance(data.observations)
+    kernel = gpflow.kernels.Matern52(variance=variance, lengthscales=0.2 * np.ones(2,))
+    gpr = gpflow.models.GPR(astuple(data), kernel, noise_variance=1e-5)
+    set_trainable(gpr.likelihood, False)
+
+    return {OBJECTIVE: trieste.models.create_model_interface({
+        "model": gpr,
+        "optimizer": gpflow.optimizers.Scipy(),
+        "optimizer_args": {"options": dict(maxiter=100)},
+    })}
+
+model = build_model(initial_data[OBJECTIVE])
```

Comment:
we shouldn't be repeating this (everything above) in every notebook. It makes it harder for users to find what they're looking for, and is a maintenance headache

Reply:
agreed - what's the solution then?

Reply:
@joelberkeley do you have something in mind? the only alternative I see is to add all of this to the intro notebook but it's already quite long with the MES extension

Reply:
most of the notebooks make basically the same gpflow GPR, so we can just make a function in the notebook utils that does this

Reply:
as for the bit above that
```
observer = mk_observer(branin, OBJECTIVE)
# I plan to make Box accept lists, so the rest will be a one-liner 
lower_bound = tf.cast(mins, gpflow.default_float())
upper_bound = tf.cast(maxs, gpflow.default_float())
search_space = trieste.space.Box(lower_bound, upper_bound)
```

Reply:
above that, we'll make plotting functions for observers, so that should be a one-liner too

Reply:
the rest is text, and we shouldn't be duplicating that, and imports